rosbag2_transport/src/rosbag2_transport/topic_filter.cpp

```diff
@@ -129,36 +149,85 @@ bool TopicFilter::take_topic(
     return false;
   }
 
-  if (!record_options_.topics.empty() && !topic_in_list(topic_name, record_options_.topics)) {
-    return false;
-  }
+  if (!is_service_event_topic) {
+    if (!record_options_.all_topics &&
+      record_options_.topics.empty() &&
+      record_options_.regex.empty() &&
+      !record_options_.include_hidden_topics)
+    {
+      return false;
+    }
 
-  std::regex exclude_regex(record_options_.exclude);
-  if (!record_options_.exclude.empty() && std::regex_search(topic_name, exclude_regex)) {
-    return false;
-  }
+    if (!record_options_.topics.empty() && !topic_in_list(topic_name, record_options_.topics)) {
+      return false;
+    }
 
-  std::regex include_regex(record_options_.regex);
-  if (
-    !record_options_.all &&  // All takes precedence over regex
-    !record_options_.regex.empty() &&  // empty regex matches nothing, but should be ignored
-    !std::regex_search(topic_name, include_regex))
-  {
-    return false;
-  }
+    if (!record_options_.exclude_topics.empty() &&
+      topic_in_list(topic_name, record_options_.exclude_topics))
+    {
+      return false;
+    }
 
-  if (!has_single_type(topic_name, topic_types)) {
-    return false;
-  }
+    std::regex include_regex(record_options_.regex);
+    if (!record_options_.all_topics &&  // All takes precedence over regex
+      !record_options_.regex.empty() &&  // empty regex matches nothing, but should be ignored
+      !std::regex_search(topic_name, include_regex))
+    {
+      return false;
+    }
 
-  if (!record_options_.include_hidden_topics && topic_is_hidden(topic_name)) {
-    RCUTILS_LOG_WARN_ONCE_NAMED(
-      ROSBAG2_TRANSPORT_PACKAGE_NAME,
-      "Hidden topics are not recorded. Enable them with --include-hidden-topics");
-    return false;
+    if (!record_options_.exclude_regex.empty()) {
+      std::regex exclude_regex(record_options_.exclude_regex);
+      if (std::regex_search(topic_name, exclude_regex)) {
+        return false;
+      }
+    }
+
+    if (!record_options_.include_hidden_topics && topic_is_hidden(topic_name)) {
+      RCUTILS_LOG_WARN_ONCE_NAMED(
+        ROSBAG2_TRANSPORT_PACKAGE_NAME,
+        "Hidden topics are not recorded. Enable them with --include-hidden-topics");
+      return false;
+    }
+  } else {
+    if (!record_options_.all_services &&
+      record_options_.services.empty() &&
+      record_options_.regex.empty())
+    {
+      return false;
+    }
+
+    if (!record_options_.services.empty() &&
+      !service_in_list(topic_name, record_options_.services))
+    {
+      return false;
+    }
+
+    if (!record_options_.exclude_services.empty() &&
+      service_in_list(topic_name, record_options_.exclude_services))
+    {
```

Comment:
@MichaelOrlov 

In the exclude services list, service event topic are used.    
The exclude service names are converted here.  
https://github.com/Barry-Xu-2018/rosbag2/blob/1c73a3f65e9b07fb08c248d0898cedc239cb5ff9/ros2bag/ros2bag/verb/record.py#L291-L292

```c++
// Convert service event topic name to service name
    auto service_name = rosbag2_cpp::service_event_topic_name_to_service_name(topic_name);
```
The following code is related to regex check. The regex string is from user who only consider service name (not service event topic). So I convert service event topic to service name and then do regex action.  

```c++
    std::regex include_regex(record_options_.regex);
    if (!record_options_.all_services &&  // All takes precedence over regex
      !record_options_.regex.empty() &&  // empty regex matches nothing, but should be ignored
      !std::regex_search(service_name, include_regex))
    {
      return false;
    }

    if (!record_options_.exclude_regex.empty()) {
      std::regex exclude_regex(record_options_.exclude_regex);
      if (std::regex_search(service_name, exclude_regex)) {
        return false;
      }
    }
```

Reply:
@Barry-Xu-2018 Ok. To avoid further confusions for others may I ask you to rename `RecordOptions::exclude_services` to the `RecordOptions::exclude_service_events` ?

Reply:
Okay. I will rename it.